core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -212,10 +222,34 @@ class ContainerProxy(
     // and we keep it in case we need to destroy it.
     case Event(data: PreWarmedData, _) => stay using data
 
-    // Run was successful
-    case Event(data: WarmedData, _) =>
+    // Init was successful
+    case Event(data: WarmedData, _: PreWarmedData) =>
+      //in case concurrency supported, multiple runs can begin as soon as init is complete
       context.parent ! NeedWork(data)
-      goto(Ready) using data
+      stay using data
+
+    // Run was successful
+    case Event(_: WarmedData, s: WarmedData) =>
```

Comment:
@tysonnorris  Here use of `WarmedData` both as message and state is bit confusing. May be we use new event type

1. For successful - `RunCompleted`
2. For init - `Init`

Further we set the `activeActivationCount` in `PreWarmedData` to 0 and make it constant i.e. not provide a way to change it from 0 for `PreWarmedData`.

Reply:
Yes, just mentioned that above - I will take a stab at replacing it with RunCompleted.